### PR TITLE
calculating Mass accretion rate for SN post-processing

### PR DESCRIPTION
**Pull request by mari2895**
## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.

### CODE REVIEW

File: src/analysis/history.cpp [hunk 1]
```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+  if (ndim >= 2) {
+    div_mass_flux_integral +=
+        -(f(b, 2, k, j + 1, i) - f(b, 2, k, j, i)) * coords.FaceArea<X2DIR>(k, j, i);
+  }
+  if (ndim >= 3) {
+    div_mass_flux_integral +=
+        -(f(b, 3, k + 1, j, i) - f(b, 3, k, j, i)) * coords.FaceArea<X3DIR>(k, j, i);
+  }
+  return div_mass_flux_integral;
+}
```
Comment (Yurlungur) [inline src/analysis/history.cpp:301]: Maybe move this into `history.hpp` so it can be re-used.

File: src/analysis/history.cpp [hunk 2]
```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
 } // exp
 
+using namespace parthenon::package::prelude;
+template <typename F>
+KOKKOS_INLINE_FUNCTION Real ComputeDivInPillbox(int ndim, int b, int k, int j, int i,
+                                                const parthenon::Coordinates_t &coords,
+                                                const F &f) {
+  Real div_mass_flux_integral;
+  div_mass_flux_integral =
+      -(f(b, 1, k, j, i + 1) - f(b, 1, k, j, i)) * coords.FaceArea<X1DIR>(k, j, i);
```
Comment (Yurlungur) [inline src/analysis/history.cpp:290]: Remove the minus signs here, and add them to the value in `mdot` as the divergence doesn't have a minus sign in its definition but `mdot` does, due to the orientation of surface normals.

File: src/analysis/history.cpp [hunk 3]
```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+
+        Real C[NDFULL];
+        geom.Coords(CellLocation::Cent, b, k, j, i, C);
+        Real r = std::sqrt(C[1] * C[1] + C[2] * C[2] + C[3] * C[3]);
+        if (r <= rc) {
+          if (gain) {
+            auto rad = pmb->packages.Get("radiation").get();
+            const parthenon::AllReduce<bool> *pdo_gain_reducer =
+                rad->MutableParam<parthenon::AllReduce<bool>>("do_gain_reducer");
+            const bool do_gain = pdo_gain_reducer->val;
```
Comment (Yurlungur) [inline src/analysis/history.cpp:346]: This can't be here---you can't access packages from within a par for on GPU. Move this above.

**Review by Yurlungur — CHANGES_REQUESTED** [review]

File: src/analysis/history.cpp (hunk 2 above)
Comment (mari2895) [reply src/analysis/history.cpp:290]: done

File: src/analysis/history.cpp (hunk 1 above)
Comment (mari2895) [reply src/analysis/history.cpp:301]: done

File: src/analysis/history.cpp [hunk 4]
```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+      resolved_pkgs.get(), {}, {PDOpt::WithFluxes});
+  auto v = desc.GetPack(md);
+
+  const int nblocks = v.GetNBlocks();
+  auto geom = Geometry::GetCoordinateSystem(md);
+  Real result = 0.0;
+  const bool is_monopole_cart =
+      (typeid(PHOEBUS_GEOMETRY) == typeid(Geometry::MonopoleCart));
+  const bool is_monopole_sph =
+      (typeid(PHOEBUS_GEOMETRY) == typeid(Geometry::MonopoleSph));
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/analysis/history.cpp:329]: done

File: src/progenitor/progenitordata.cpp [hunk 5]
```diff
@@ -116,14 +116,25 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
+    return ReduceInGain<class fluid_cons::density>(md, 1, 0);
   };
   auto Qgain = [](MeshData<Real> *md) {
-    return ReduceInGain<class internal_variables::GcovHeat>(md, 0) -
-           ReduceInGain<class internal_variables::GcovCool>(md, 0);
+    return ReduceInGain<class internal_variables::GcovHeat>(md, 0, 0) -
+           ReduceInGain<class internal_variables::GcovCool>(md, 0, 0);
+  };
+  auto Mdot400 = [](MeshData<Real> *md) {
+    Real rc = 0.04;
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/progenitor/progenitordata.cpp:126]: I am trying to imagine how is the second part going to work

File: src/analysis/history.cpp (hunk 3 above)
Comment (mari2895) [reply src/analysis/history.cpp]: done

File: src/analysis/history.cpp [hunk 6]
```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+            auto rad = pmb->packages.Get("radiation").get();
+            const parthenon::AllReduce<bool> *pdo_gain_reducer =
+                rad->MutableParam<parthenon::AllReduce<bool>>("do_gain_reducer");
+            const bool do_gain = pdo_gain_reducer->val;
+            bool is_netheat = (v(b, internal_variables::GcovHeat(), k, j, i) -
+                                   v(b, internal_variables::GcovCool(), k, j, i) >
+                               1.e-8); // checks that in the gain region
+            auto analysis = pmb->packages.Get("analysis").get();
+            const Real inside_pns_threshold =
+                analysis->Param<Real>("inside_pns_threshold");
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/analysis/history.cpp]: done

File: src/analysis/history.hpp [hunk 7]
```diff
@@ -112,11 +115,20 @@ Real ReduceInGain(MeshData<Real> *md, int idx = 0) {
       kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
       KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
         Real gdet = geom.DetGamma(CellLocation::Cent, 0, k, j, i);
-        bool is_netheat =
-            (v(b, iv::GcovHeat(), k, j, i) - v(b, iv::GcovCool(), k, j, i) > 0);
+        bool is_netheat = (v(b, iv::GcovHeat(), k, j, i) - v(b, iv::GcovCool(), k, j, i) >
+                           1.e-8); // checks that in the gain region
+        auto analysis = pmb->packages.Get("analysis").get();
+        const Real outside_pns_threshold = analysis->Param<Real>("outside_pns_threshold");
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/analysis/history.hpp]: done

File: src/analysis/history.hpp [hunk 8]
```diff
@@ -112,11 +115,20 @@ Real ReduceInGain(MeshData<Real> *md, int idx = 0) {
...
+        bool is_outside_pns = (v(b, fluid_prim::entropy(), k, j, i) >
+                               outside_pns_threshold); // checks that outside PNS
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/analysis/history.hpp:123]: It seems to be working so let's keep for now maybe

File: src/progenitor/progenitordata.cpp [hunk 9]
```diff
@@ -116,14 +116,25 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
   parthenon::HstVar_list hst_vars = {};
   auto Mgain = [](MeshData<Real> *md) {
-    return ReduceInGain<class fluid_prim::density>(md, 0);
+    return ReduceInGain<class fluid_cons::density>(md, 1, 0);
   };
   auto Qgain = [](MeshData<Real> *md) {
-    return ReduceInGain<class internal_variables::GcovHeat>(md, 0) -
-           ReduceInGain<class internal_variables::GcovCool>(md, 0);
+    return ReduceInGain<class internal_variables::GcovHeat>(md, 0, 0) -
+           ReduceInGain<class internal_variables::GcovCool>(md, 0, 0);
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/progenitor/progenitordata.cpp]: deleted and still compiles and runs so is not required.

File: src/analysis/analysis.cpp [hunk 10]
```diff
@@ -6,10 +6,20 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   auto analysis_pkg = std::make_shared<StateDescriptor>("analysis");
   Params &params = analysis_pkg->AllParams();
   Real sigma = pin->GetOrAddReal("analysis", "sigma", 2e-3);
+  Real outside_pns_threshold =
+      pin->GetOrAddReal("analysis", "outside_pns_threshold",
+                        2.42e-5); // corresponds to entropy > 3 kb/baryon
+  Real inside_pns_threshold = pin->GetOrAddReal("analysis", "inside_pns_threshold",
+                                                0.008); // corresponds to r < 80 km
+  Real radius_cutoff_mdot =
+      pin->GetOrAddReal("analysis", "radius_cutoff_mdot", 0.04); // default 400km
```
Comment (Yurlungur) [inline src/analysis/analysis.cpp:15]: This stuff might better belong in the progenitor package.

File: src/progenitor/progenitordata.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yurlungur) [reply src/progenitor/progenitordata.cpp]: `ParameterInput` can read a `std:::vector` so you could do something like this:
```C++
auto mdot_radii = pin->GetOrAddVector("analysis", "mdot_radii", std::vector<Real>{0.04});
for (auto rc : mdot_radii) {
  auto mdot = [=](MeshData<Real> *md) { return History::CalculateMdot(md, rc, 0); };
  hst_vars.emplace_back(HistoryOutputVar(HstSum, mdot, "Mdot at r = " + std::to_string(rc)));
}
```

File: src/analysis/analysis.cpp (hunk 10 above)
Comment (mari2895) [reply src/analysis/analysis.cpp]: moved to progenitor pkg

File: src/analysis/history.cpp [hunk 11]
```diff
@@ -279,4 +280,75 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+        const Real vol = coords.CellVolume(k, j, i);
+
+        Real C[NDFULL];
+        geom.Coords(CellLocation::Cent, b, k, j, i, C);
+        Real r = std::sqrt(C[1] * C[1] + C[2] * C[2] + C[3] * C[3]);
+        if (r <= rc) {
+          if (gain) {
+            bool is_netheat = (v(b, internal_variables::GcovHeat(), k, j, i) -
+                                   v(b, internal_variables::GcovCool(), k, j, i) >
+                               1.e-8); // checks that in the gain region
```
Comment (Yurlungur) [inline src/analysis/history.cpp:327]: Maybe make the threshold a runtime parameter?

File: src/analysis/history.hpp [hunk 12]
```diff
@@ -105,23 +108,52 @@ Real ReduceInGain(MeshData<Real> *md, int idx = 0) {
...
   parthenon::par_reduce(
       parthenon::LoopPatternMDRange(),
       "Phoebus History for integrals in gain region (SN)", DevExecSpace(), 0, nblocks - 1,
       kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
       KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
         Real gdet = geom.DetGamma(CellLocation::Cent, 0, k, j, i);
-        bool is_netheat =
-            (v(b, iv::GcovHeat(), k, j, i) - v(b, iv::GcovCool(), k, j, i) > 0);
+        bool is_netheat = (v(b, iv::GcovHeat(), k, j, i) - v(b, iv::GcovCool(), k, j, i) >
+                           1.e-8); // checks that in the gain region
```
Comment (Yurlungur) [inline src/analysis/history.hpp:121]: See above

File: src/analysis/history.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Yurlungur) [reply src/analysis/history.hpp:123]: ok

File: src/progenitor/progenitordata.cpp [hunk 13]
```diff
@@ -83,6 +84,18 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
+  Real outside_pns_threshold =
+      pin->GetOrAddReal("progenitor", "outside_pns_threshold",
+                        2.42e-5); // corresponds to entropy > 3 kb/baryon
+  Real inside_pns_threshold = pin->GetOrAddReal("progenitor", "inside_pns_threshold",
+                                                0.008); // corresponds to r < 80 km
+  UnitConversions units(pin);
+  Real LengthCGSToCode = units.GetLengthCGSToCode();
+  Real rc = 400e5 * LengthCGSToCode;
+  auto mdot_radii = pin->GetOrAddVector("progenitor", "mdot_radii",
+                                        std::vector<Real>{rc}); // default 400km
```
Comment (Yurlungur) [inline src/progenitor/progenitordata.cpp:97]: I don't think you want to conversion here.
```suggestion
  Real LengthCGSToCode = units.GetLengthCGSToCode();
  auto mdot_radii = pin->GetOrAddVector("progenitor", "mdot_radii",
                                        std::vector<Real>{400}); // default 400km
```

File: src/progenitor/progenitordata.cpp [hunk 14]
```diff
@@ -116,14 +133,28 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
   };
   auto Qgain = [](MeshData<Real> *md) {
-    return ReduceInGain<class internal_variables::GcovHeat>(md, 0) -
-           ReduceInGain<class internal_variables::GcovCool>(md, 0);
+    return ReduceInGain<internal_variables::GcovHeat>(md, 0, 0) -
+           ReduceInGain<internal_variables::GcovCool>(md, 0, 0);
+  };
+  for (auto rc : mdot_radii) {
+    auto Mdot = [rc, LengthCGSToCode](MeshData<Real> *md) {
+      return History::CalculateMdot(md, rc, 0);
```
Comment (Yurlungur) [inline src/progenitor/progenitordata.cpp:144]:
```suggestion
    auto rc_code = rc*1e5*LengthCGSToCode;
    auto Mdot = [rc_code](MeshData<Real> *md) {
      return History::CalculateMdot(md, rc_code, 0);
```

File: src/progenitor/progenitordata.cpp [hunk 15]
```diff
@@ -116,14 +133,28 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
+    };
+    hst_vars.emplace_back(HistoryOutputVar(
+        HstSum, Mdot,
+        "Mdot at r = " + std::to_string(int(rc / LengthCGSToCode / 1.e5)) + "km"));
```
Comment (Yurlungur) [inline src/progenitor/progenitordata.cpp:148]: With above, you can remove the conversion here. However, I don't think you should cast to int. What if you choose 400.5 km for some reason?
```suggestion
        "Mdot at r = " + std::to_string(rc) + "km"));
```

File: src/analysis/history.cpp (hunk 11 above)
Comment (mari2895) [reply src/analysis/history.cpp:327]: in analysis pkg or progenitor?

Comment (Yurlungur) [reply src/analysis/history.cpp:327]: progenitor... probably.

File: src/progenitor/progenitordata.cpp (hunk 15 above)
Comment (Yurlungur) [reply src/progenitor/progenitordata.cpp:148]: On second thought, it's fine to cast to an int here... probably nobody will want to look at half-km here.

Comment (mari2895) [reply src/progenitor/progenitordata.cpp:148]: yes exactly

File: src/analysis/history.cpp [hunk 16]
```diff
@@ -279,4 +280,75 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
...
+        const auto &coords = v.GetCoordinates(b);
+        const Real vol = coords.CellVolume(k, j, i);
+
+        Real C[NDFULL];
+        geom.Coords(CellLocation::Cent, b, k, j, i, C);
+        Real r = std::sqrt(C[1] * C[1] + C[2] * C[2] + C[3] * C[3]);
+        if (r <= rc) {
+          if (gain) {
+            bool is_netheat = (v(b, internal_variables::GcovHeat(), k, j, i) -
+                                   v(b, internal_variables::GcovCool(), k, j, i) >
```
[the remark this answers is not in the document]
Comment (mari2895) [reply src/analysis/history.cpp:326]: done

File: src/analysis/history.cpp (hunk 11 above)
Comment (mari2895) [reply src/analysis/history.cpp:327]: done

File: src/analysis/history.hpp (hunk 12 above)
Comment (mari2895) [reply src/analysis/history.hpp:121]: done

**Review by Yurlungur — APPROVED** [review]
As soon as tests pass here (and you confirm it works for you locally) I will merge.

**Comment by mari2895** [comment]
> As soon as tests pass here (and you confirm it works for you locally) I will merge.

All tests have passed, compiles, runs, produces reasonable output.

**Pull request merged**